staging/src/k8s.io/apiserver/pkg/endpoints/filters/storageversion.go

```diff
@@ -69,6 +73,31 @@ func WithStorageVersionPrecondition(handler http.Handler, svm storageversion.Man
 			handler.ServeHTTP(w, req)
 			return
 		}
+		// The system namespace is required for apiserver-identity lease to exist. Allow the apiserver
+		// itself to create namespaces.
+		// NOTE: with this exception, if the bootstrap client writes namespaces with a new version,
+		// and the upgraded apiserver dies before updating the StorageVersion for namespaces, the
+		// storage migrator won't be able to tell these namespaces are stored in a different version in etcd.
+		// Because the bootstrap client only creates system namespace and doesn't update them. This can
+		// only happen if the upgraded apiserver is the first apiserver that kicks off namespace creation,
+		// or if a upgraded server that joins an existing cluster have new system namespaces (other
+		// than kube-system, kube-public, kube-node-lease) that need to be created.
```

Comment:
kube-system namespace is only created when the cluster is newly created. So the etcd corruption only happens if the cluster admin creates a cluster, the kube-system namespace is created, but the cluster creation fails, then the cluster admin creates a cluster of a different version which supports a different encoding version for namespaces, but keeps using the same etcd.

I think this is an extremely rare use case.

Reply:
I agree rolling back a just-created-and-crashed cluster without recreating etcd is weird.

The other case is possible. Adding a new kube-reserved namespace to an existing cluster could leave object with wrong encoding version in etcd.

Reply:
Ack.

Can we look into the body of the request to get the name of the namespace? We can explore in a followup.

Reply:
I think https://github.com/kubernetes/kubernetes/pull/95978#discussion_r522391599 is required for both:
- solving this leak
- providing way to recover a stuck cluster (not sure if the use case is bootstrap only, will think about it later)

I'm not sure if deserializing the body in this filter is achievable. Could be a faster workaround if it is not hard to do. Will explore